pkg/ingester/incremental_transfer.go

```diff
@@ -0,0 +1,260 @@
+package ingester
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/weaveworks/common/user"
+	"golang.org/x/net/context"
+)
+
+var (
+	blockedRanges = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "cortex_ingester_blocked_ranges",
+		Help: "The current number of ranges that will not accept writes by this ingester.",
+	})
+)
+
+func init() {
+	prometheus.MustRegister(blockedRanges)
+}
+
+// TransferChunksSubset accepts chunks from a client and moves them into the local Ingester.
+func (i *Ingester) TransferChunksSubset(stream client.Ingester_TransferChunksSubsetServer) error {
+	i.userStatesMtx.Lock()
+	defer i.userStatesMtx.Unlock()
+
+	fromIngesterID, seriesReceived, err := i.acceptChunksFromStream(acceptChunksOptions{
+		States: i.userStates,
+		Stream: stream,
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := stream.SendAndClose(&client.TransferChunksResponse{}); err != nil {
+		level.Error(util.Logger).Log("msg", "Error closing SendChunks stream", "from_ingester", fromIngesterID, "err", err)
+		return err
+	}
+
+	// Target ingester may not have had any streams to send.
+	if seriesReceived == 0 {
+		return nil
+	} else if fromIngesterID == "" {
+		level.Error(util.Logger).Log("msg", "received TransferChunks request with no ID from ingester")
+		return fmt.Errorf("no ingester id")
+	}
+
+	return nil
+}
+
+// GetChunksSubset accepts a get request from a client and sends all chunks from the serving ingester
+// that fall within the given range to the client.
+func (i *Ingester) GetChunksSubset(req *client.GetChunksRequest, stream client.Ingester_GetChunksSubsetServer) error {
+	userStatesCopy := i.userStates.cp()
+	if len(userStatesCopy) == 0 {
+		level.Info(util.Logger).Log("msg", "nothing to transfer")
+		return nil
+	}
+
+	// Block the ranges locally so we don't receive traffic for it anymore. The caller should decide
+	// when ranges should be unblocked: we should continue to reject writes for as long as we may
+	// receive them. When the joining token has been completely inserted into the ring, it will
+	// be safe to remove the blocks.
+	i.BlockRanges(req.Ranges)
```

Comment:
shouldn't the block occur before copying user states in the unlikely case there's a new series appended during `i.userStates.cp()`?

Reply:
This shouldn't happen because the new token isn't in the ring yet, but it's better to be safe here; I'll move the block.